lib/licensed/sources/nuget.rb

```diff
@@ -0,0 +1,230 @@
+            return if redirect_limit == 0
+
+            begin
+              response = Net::HTTP.get_response(url)
+              case response
+              when Net::HTTPSuccess     then
+                @response_by_url[url] = response.body
+              when Net::HTTPRedirection then
+                redirect_url = URI.parse(response["location"])
+                if redirect_url.relative?
+                  redirect_url = url + redirect_url
+                end
+                # The redirect might be to a URL that requires transformation, i.e. a github file
+                redirect_url = text_content_url(redirect_url.to_s)
+                @response_by_url[url] = fetch_content(redirect_url, redirect_limit - 1)
+              end
+            rescue => error
+              # Host might no longer exist or some other error, ignore
+            end
+          end
+        end
+      end
+
+      def enabled?
+        nuget_config_exists(config.pwd)
+      end
+
+      def nuget_obj_root
+        config.dig("nuget", "obj_root") || config.pwd
+      end
+
+      def dump_projects?
+        config.dig("nuget", "projects", "dump")
+      end
+
+      def exclude_projects
+        config.dig("nuget", "projects", "exclude") || []
+      end
+
+      def excluded_project?(project_name)
+        exclude_projects.any? do |pattern|
+          File.fnmatch?(pattern, project_name, File::FNM_PATHNAME | File::FNM_CASEFOLD)
+        end
+      end
+
+      def nuget_config_exists(root)
+        # Multiple supported casings: https://github.com/NuGet/Home/issues/1427
+        File.exist?(File.join(root, "nuget.config")) || File.exist?(File.join(root, "NuGet.config")) || File.exist?(File.join(root, "NuGet.Config"))
+      end
+
+      def enumerate_dependencies
+        packages.map do |key, package|
+          metadata = {
+            "type" => NuGet.type,
+            "name" => package["name"]
+          }
+
+          # Emit the names of the projects that consume a particular dependency.
+          # Useful for determining projects to exclude.
+          metadata["projects"] = package["projects"].to_a.sort if dump_projects?
+
+          NuGetDependency.new(
+            name: "#{package["name"]}-#{package["version"]}",
+            version: package["version"],
+            path: package["path"],
+            metadata: metadata
+          )
+        end
+      end
+
+      # Inspect project.assets.json files for package references.
+      # Ideally we'd use `dotnet list package` instead, but its output isn't
+      # easily machine readable, and it also requires repos to have a .sln file
+      # to evaluate multiple projects.
+      def packages
+        return @packages_by_id if defined?(@packages_by_id)
+        @packages_by_id = Hash.new
+        Dir.glob(File.join(nuget_obj_root, "**/project.assets.json")).map do |file|
+          gather_packages(file)
+        end
+        @packages_by_id
+      end
+
+      def gather_packages(project_assets_file)
```

Comment:
A few ruby idioms on this one 👍 
1. It's preferable to pass state into functions rather than mutate an instance variable.
2. `each` is preferable to `map` in this case since you're not using the returned array from `map` 
3. You can clean this up by using `each_with_object` so you don't need to predefine `@package_by_id`

```suggestion
      def packages
        return @packages_by_id if defined?(@packages_by_id)
        search_path = File.join(nuget_obj_root, "**/project.assets.json")
        @packages_by_id = Dir.glob(search_path).each_with_object({}) do |file, packages|
          gather_packages(file, packages)
        end
      end

      def gather_packages(project_assets_file, packages)
```

With further updates to `gather_packages` to use the passed in hash rather than the instance variable.

Reply:
Thanks, was not aware of the use case for `each_with_object`, now it makes sense. Will fix up gather_packages.